### ISSUE TITLE
Snapping ScrollPhysics

### ISSUE THREAD

**Issue by timrijckaert** [flutter/flutter#41472]
Hi, 

After some tinkering and Googling all over the place I became super frustrated with the API / lack of documentation for `ScrollPhysics`. 
On Android you can use what's called a [SnapHelper](https://developer.android.com/reference/android/support/v7/widget/LinearSnapHelper) inside your RecyclerView (analogous to a ListView in Flutter) that will automatically snap to a certain position.

The SnapHelper does this on a position based API.  
You can ask which View is currently in your chosen ViewPort and get its position and ask the RecyclerView to animate to that position.

Flutter on the other hand wants us to work with logical pixels, which makes this super trivial, common pattern difficult to implement.  
All the _solutions_ I found was to use items inside the list that have a fixed width/height and don't account for flinging gestures.

![a92ca-1npi-wjo0ovasbkwv0gy5rw](https://user-images.githubusercontent.com/6739443/65771459-86f2fb80-e138-11e9-9eaf-9a9dd3bb0e9d.gif)

tl;dr How to implement this 👆in Flutter so it works for any item in a ListView?

**Comment by timrijckaert** [comment]
I'm including the poor mans version we are currently using.
Which works, but not like we are used to on Android.

Especially passing the itemWidth is an eyesore

```
class SnappingListScrollPhysics extends ScrollPhysics {
  final double itemWidth;

  const SnappingListScrollPhysics({
    @required this.itemWidth,
    ScrollPhysics parent,
  }) : super(parent: parent);

  @override
  SnappingListScrollPhysics applyTo(ScrollPhysics ancestor) => SnappingListScrollPhysics(
        parent: buildParent(ancestor),
        itemWidth: itemWidth,
      );

  double _getItem(ScrollPosition position) => (position.pixels) / itemWidth;

  double _getPixels(ScrollPosition position, double item) => min(item * itemWidth, position.maxScrollExtent);

  double _getTargetPixels(ScrollPosition position, Tolerance tolerance, double velocity) {
    double item = _getItem(position);
    if (velocity < -tolerance.velocity) {
      item -= 0.5;
    } else if (velocity > tolerance.velocity) {
      item += 0.5;
    }
    return _getPixels(position, item.roundToDouble());
  }

  @override
  Simulation createBallisticSimulation(ScrollMetrics position, double velocity) {
    // If we're out of range and not headed back in range, defer to the parent
    // ballistics, which should put us back in range at a page boundary.
    if ((velocity <= 0.0 && position.pixels <= position.minScrollExtent) ||
        (velocity >= 0.0 && position.pixels >= position.maxScrollExtent)) {
      return super.createBallisticSimulation(position, velocity);
    }
    final Tolerance tolerance = this.tolerance;
    final double target = _getTargetPixels(position, tolerance, velocity);
    if (target != position.pixels) {
      return ScrollSpringSimulation(spring, position.pixels, target, velocity, tolerance: tolerance);
    }
    return null;
  }

  @override
  bool get allowImplicitScrolling => false;
}
```

**Comment by janmoppel** [comment]
Hello @timrijckaert !

Thank you for posting.
Your question seems to be a good fit for the [StackOverflow](https://stackoverflow.com/tags/flutter) as well. Flutter has a very active and helpful community there, so feel free to post it there too.

**Comment by timrijckaert** [comment]
I did, but this is the place where you'll found the master minds that work @Google.
Either way I cross posted it see [StackOverflow](https://stackoverflow.com/questions/58135446/snapping-behavior-custom-scrollphysics?noredirect=1#comment102664465_58135446)

**Comment by RomanJos** [comment]
Here is my version of the one from @timekone, it extends the IOS one and offer some improvements.

I made some adjustements in the target predictions, I divide the velocity by 1000 to predict the next item to scroll, you can divide further if you want more friction.
The only downside is that it doesn't behave exactly like IOS when you scroll from the middle for example, with a lot of velocity, it will end at the min/max extent instead of going a little bit further and going back.
You can modify _getPixels() like this to have a similar effect but it doens't look so good :
`min(max(item * itemWidth, position.minScrollExtent - 35), position.maxScrollExtent + 35);`


```dart
class SnapBounceScrollPhysics extends BouncingScrollPhysics {
  final double itemWidth;
  const SnapBounceScrollPhysics({
    @required this.itemWidth,
    ScrollPhysics parent,
  }) : super(parent: parent);

  @override
  SnapBounceScrollPhysics applyTo(ScrollPhysics ancestor) => SnapBounceScrollPhysics(
        parent: buildParent(ancestor),
        itemWidth: itemWidth,
      );

  double _getItem(ScrollPosition position) => (position.pixels) / itemWidth;

  double _getPixels(ScrollPosition position, double item) =>
      min(max(item * itemWidth, position.minScrollExtent), position.maxScrollExtent);

  double _getTargetPixels(ScrollPosition position, Tolerance tolerance, double velocity) {
    double item = _getItem(position);
    item += velocity / 1000;
    return _getPixels(position, item.roundToDouble());
  }

  @override
  Simulation createBallisticSimulation(ScrollMetrics position, double velocity) {
    // If we're out of range and not headed back in range, defer to the parent
    // ballistics, which should put us back in range at a page boundary.
    if ((velocity <= 0.0 && position.pixels <= position.minScrollExtent) ||
        (velocity >= 0.0 && position.pixels >= position.maxScrollExtent)) {
      return super.createBallisticSimulation(position, velocity);
    }
    final Tolerance tolerance = this.tolerance;
    final double target = _getTargetPixels(position, tolerance, velocity);
    if (target != position.pixels) {
      return ScrollSpringSimulation(spring, position.pixels, target, velocity, tolerance: tolerance);
    }
    return null;
  }

  @override
  bool get allowImplicitScrolling => false;
}
 ```

**Comment by jamesblasco** [comment]
Hello there! Is this not possible to do this with  [FixedExtentScrollPhysics](https://api.flutter.dev/flutter/widgets/FixedExtentScrollPhysics-class.html)

**Comment by polarby** [comment]
@RomanJos, @timrijckaert Any idea of how to implement this with dynamic heights/widths? (Passing the heights in as a parameter list)

**Comment by RomanJos** [comment]
@polarby I believe having a list where each child enter its size in its index (using performLayout in the childs or using a wrapper that has it I believe, its been a while I haven't used Flutter) with the ballistic simulation looking at it instead of blindly doing `index * itemWidth` to calculate the snap point.

The childs feed the list on each layout change and the scroll things use it in its calculation.

Hope it helped
